cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -27,20 +27,135 @@ This file proves useful properties of canonical list-based maps defined in
 
 namespace Cedar.Data.Map
 
-/-! ### Well-formed sets -/
+/-! ### Well-formed maps -/
 
 def WellFormed {α β} [LT α] [DecidableLT α] (m : Map α β) :=
   m = Map.make m.toList
 
-/-! ### contains and mem -/
+/--
+  In well-formed maps, if there are two pairs with the same key, then they have
+  the same value
+-/
+theorem key_maps_to_one_value [BEq α] [LT α] [DecidableLT α] (k : α) (v₁ v₂ : β) (m : Map α β) :
+  m.WellFormed →
+  (k, v₁) ∈ m.kvs →
+  (k, v₂) ∈ m.kvs →
+  v₁ = v₂
+:= by
+  unfold WellFormed toList
+  intro wf h₁ h₂
+  sorry
+
+/--
+  In well-formed maps, if a key maps to a value, then all equal keys (by BEq)
+  also map to that value
+-/
+theorem equal_keys_same_value [BEq α] [LT α] [DecidableLT α] (k₁ k₂ : α) (v : β) (m : Map α β) :
+  m.WellFormed →
+  (k₁, v) ∈ m.kvs →
+  k₁ == k₂ →
+  (k₂, v) ∈ m.kvs
+:= by
+  unfold WellFormed toList
+  intro wf h₁ h₂
+  sorry
+
+/--
+  `make` on equivalent lists of (k,v) pairs produces equal maps
+
+  (Organizationally belongs in the `make` section, but is needed before that)
+-/
+theorem make_eqv [LT α] [DecidableLT α] {xs ys : List (α × β)}:
+  xs ≡ ys →
+  Map.make xs = Map.make ys
+:= by
+  unfold make
+  intro h₁
+  sorry
+
+/--
+  If two well-formed maps have equivalent (k,v) sets, then the maps are actually
+  equal
+-/
+theorem eq_iff_kvs_equiv [LT α] [DecidableLT α] {m₁ m₂ : Map α β}
+  (wf₁ : m₁.WellFormed)
+  (wf₂ : m₂.WellFormed) :
+  m₁.kvs ≡ m₂.kvs ↔ m₁ = m₂
+:= by
+  constructor
+  case mp =>
+    intro h₁
+    unfold WellFormed toList at wf₁ wf₂
+    rw [wf₁]
+    rw [wf₂]
+    exact make_eqv h₁
+  case mpr =>
+    intro h₁
+    subst h₁
+    exact List.Equiv.refl
+
+/-! ### contains, mem, and find? -/
 
-theorem in_list_in_map {α : Type u} (k : α) (v : β) (m : Map α β) :
+theorem in_list_in_map {α : Type u} {k : α} {v : β} {m : Map α β} :
```

Comment:
Emina, I thought you would (like me) be bugged by how inconsistent `Thm/Data/Map.lean` currently is about whether these `k`, `v`, and `m` arguments are `{}` or not.  (Compare to `contains_iff_some_find?`, immediately below this theorem.)

Adapting to this change is extremely trivial.  In most cases you can just remove the explicit `k`, `v`, and `m` arguments from callsites, because Lean can infer these arguments from the hypothesis argument.  If not, you can just add `@` and it will work exactly the same as before.

I think the consistency is worth the minor pain of updating usage sites of this theorem.

Reply:
Yeah, it's definitely bugging me, and we should fix it :).  

I'm hoping we can do that as part of a separate future "breaking changes" PR where all / most breaking changes we want to make are included.  (We talked about a few others offline---e.g., `toList` vs `kvs` and `elts` for sets.  I think you also also wanted to change how Map membership is defined, which is a good idea too.)

Handling them incrementally is more disruptive, at least on my end.

Should we create an issue on this repo to track all the changes we want to make and put a deadline on it?

Reply:
Separate issue sounds good.  However, I think we should do it sooner than later, perhaps before merging this PR itself, so as not to accumulate more tech debt (more usage sites that might need changes).  I can split off these changes into a separate PR for discussion, hopefully today.

I disagree that we should have one big "breaking changes" PR including all / most breaking changes.  Focused PRs are much easier to review.  If you want to batch all the breaking changes on your end, you can delay applying the focused PRs until all of them are merged, and then handle them all at once.  But I'd imagine your job would be easier too if you handled one change at a time, rather than many unrelated changes concurrently.

Reply:
In the meantime I'll revert these in favor of the separate PR